guidelines.rst

```diff
@@ -309,3 +404,193 @@ These objects should be type-checked as if they were ``PyObject*``.
+
+      *  return ``PyStatus``, or
+      *  return ``-1``/``NULL`` to signal failure, but have an alternate way
+         of reporting error details.
+
+In cases where ``-1`` or ``NULL`` is a valid result, use an
+:ref:`output argument <output argument>` to provide that result.
+
+See `return schemes`_ for concrete examples.
+
+.. note::
+
+    For background and discussions, see:
+
+    - https://github.com/capi-workgroup/api-evolution/issues/13
+    - https://github.com/capi-workgroup/decisions/issues/19
+
+
+Exceptions for infallible functions
+-----------------------------------
+
+Some functions can not fail, and callers cannot check for exceptions:
+
+* Deallocators and reference sinks like ``PyMem_Free`` and ``Py_DECREF``,
+  which use ``void`` as the return type.
+
+Other functions cannot fail, and users may *optionally* skip error
+checking:
+
+* Refcounting operations, like ``Py_NewRef``.
+* Operations on native types that cannot have exceptional cases
+  (e.g. overflow), like ``Py_HashPointer``.
+* Subtype-checking functions, like ``PyTuple_Check`` and ``PyTuple_CheckExact``,
+  which return either ``0`` or ``1``.
+  (This does not extend to *subclass* checking, like ``PyObject_IsSubclass``,
+  which can call Python code.)
+
+Even in these cases, ``-1`` and ``NULL`` are reserved for errors;
+infallible functions must never return these values.
+(This allows auto-generated API wrappers to avoid unnecessary special cases.)
+
+As always, other exceptions can be added here with approval from the
+C API working group.
+
+Mind that infallibility is very often an implementation detail
+that should not be exposed in the API. That is, some functions'
+*current CPython implementations* cannot fail,
+but we may want the function to fail (or warn) in the future.
+Users should only skip error checking if the function's documentation
+explicitly allows it.
+
+
+.. _output argument:
+
+Output arguments
+================
+
+.. (There's nothing to say for *arguments* in general, it's all under
+   types or reference conting.
+   If that changes, "Output arguments" should be a subsection of "Arguments")
+
+Output arguments are pointers to memory that a function fills in.
+Use these when a result from a function cannot be returned as the return value,
+for example:
+
+* ``-1`` or ``NULL`` is a valid (non-exceptional) result: for example,
+  in ``PyDict_GetItemRef``.
+* There are multiple results: for example, in ``PyUnicode_AsUTF8AndSize``.
+
+Guidelines for output arguments:
+
+* Functions must always fill in the output arguments. If an error
+  occurs or the rseult is not available, the output should typically be set
+  to NULL or zero.
+* When it might be useful for users to call a function but ignore an output,
+  allow passing NULL as the output argument.
+* Ownership of a ``PyObject*`` result is transferred to the caller,
+  as with return values.
+  [TODO: Link to guidelines about ownership & borrowing, when those are added]
```

Comment:
```suggestion

    .. TODO: Link to guidelines about ownership & borrowing, when those are added
```

Reply:
I need to do a formatting pass before turning this into a PEP. Before that, I prefer visible TODO reminders.